src/librustc/metadata/decoder.rs

```diff
@@ -234,22 +233,25 @@ fn variant_disr_val(d: rbml::Doc) -> Option<ty::Disr> {
 
 fn doc_type<'tcx>(doc: rbml::Doc, tcx: &ty::ctxt<'tcx>, cdata: Cmd) -> Ty<'tcx> {
     let tp = reader::get_doc(doc, tag_items_data_item_type);
-    parse_ty_data(tp.data, cdata.cnum, tp.start, tcx,
-                  |_, did| translate_def_id(cdata, did))
+    TyDecoder::with_doc(tcx, cdata.cnum, tp,
+                        &mut |_, did| translate_def_id(cdata, did))
```

Comment:
@eddyb hmm perhaps -- it's a bit tricky to get lifetime of the closure to work out I think, but I guess now that the closure is stored in a struct, we could give ownership of it instead of passing an `&mut`. Would be kind of nicer anyway. Maybe I should do that.